File: storm-java21/src/main/java/st/orm/repository/impl/EntityRepositoryImpl.java
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.repository.impl;

import static java.util.Objects.requireNonNull;
import static st.orm.template.impl.StringTemplates.convert;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;
import st.orm.Data;
import st.orm.Entity;
import st.orm.Metamodel;
import st.orm.Page;
import st.orm.Pageable;
import st.orm.Ref;
import st.orm.repository.EntityRepository;
import st.orm.template.Model;
import st.orm.template.ORMTemplate;
import st.orm.template.QueryBuilder;
import st.orm.template.impl.ModelImpl;
import st.orm.template.impl.ORMTemplateImpl;
import st.orm.template.impl.QueryBuilderImpl;

/**
 */
public final class EntityRepositoryImpl<E extends Entity<ID>, ID> implements EntityRepository<E, ID> {
    private final st.orm.core.repository.EntityRepository<E, ID> core;

    public EntityRepositoryImpl(st.orm.core.repository.EntityRepository<E, ID> core) {
        this.core = requireNonNull(core);
    }

    @Override
    public Model<E, ID> model() {
        return new ModelImpl<>((st.orm.core.template.impl.ModelImpl<E, ID>) core.model());
    }

    @Override
    public Ref<E> ref(ID id) {
        return core.ref(id);
    }

    @Override
    public Ref<E> ref(E entity) {
        return core.ref(entity);
    }

    @Override
    public Ref<E> unload(E entity) {
        return core.unload(entity);
    }

    @Override
    public QueryBuilder<E, E, ID> select() {
        return new QueryBuilderImpl<>(core.select());
    }

    @Override
    public QueryBuilder<E, Long, ID> selectCount() {
        return new QueryBuilderImpl<>(core.selectCount());
    }

    @Override
    public <R> QueryBuilder<E, R, ID> select(Class<R> selectType) {
        return new QueryBuilderImpl<>(core.select(selectType));
    }

    @Override
    public QueryBuilder<E, Ref<E>, ID> selectRef() {
        return new QueryBuilderImpl<>(core.selectRef());
    }

    @Override
    public <R> QueryBuilder<E, R, ID> select(Class<R> selectType, StringTemplate template) {
        return new QueryBuilderImpl<>(core.select(selectType, convert(template)));
    }

    @Override
    public <R extends Data> QueryBuilder<E, Ref<R>, ID> selectRef(Class<R> refType) {
        return new QueryBuilderImpl<>(core.selectRef(refType));
    }

    @Override
    public QueryBuilder<E, ?, ID> delete() {
        return new QueryBuilderImpl<>(core.delete());
    }

    @Override
    public long count() {
        return core.count();
    }

    @Override
    public boolean exists() {
        return core.exists();
    }

    @Override
    public boolean existsById(ID id) {
        return core.existsById(id);
    }

    @Override
    public boolean existsByRef(Ref<E> ref) {
        return core.existsByRef(ref);
    }

    @Override
    public void insert(E entity) {
        core.insert(entity);
    }

    @Override
    public void insert(E entity, boolean ignoreAutoGenerate) {
        core.insert(entity, ignoreAutoGenerate);
    }

    @Override
    public ID insertAndFetchId(E entity) {
        return core.insertAndFetchId(entity);
    }

    @Override
    public E insertAndFetch(E entity) {
        return core.insertAndFetch(entity);
    }

    @Override
    public void update(E entity) {
        core.update(entity);
    }

    @Override
    public E updateAndFetch(E entity) {
        return core.updateAndFetch(entity);
    }

    @Override
    public void upsert(E entity) {
        core.upsert(entity);
    }

    @Override
    public ID upsertAndFetchId(E entity) {
        return core.upsertAndFetchId(entity);
    }

    @Override
    public E upsertAndFetch(E entity) {
        return core.upsertAndFetch(entity);
    }

    @Override
    public void remove(E entity) {
        core.remove(entity);
    }

    @Override
    public void removeById(ID id) {
        core.removeById(id);
    }

    @Override
    public void removeByRef(Ref<E> ref) {
        core.removeByRef(ref);
    }

    @Override
    public void removeAll() {
        core.removeAll();
    }

    @Override
    public Optional<E> findById(ID id) {
        return core.findById(id);
    }

    @Override
    public Optional<E> findByRef(Ref<E> ref) {
        return core.findByRef(ref);
    }

    @Override
    public E getById(ID id) {
        return core.getById(id);
    }

    @Override
    public E getByRef(Ref<E> ref) {
        return core.getByRef(ref);
    }

    @Override
    public <V> Optional<E> findBy(Metamodel.Key<E, V> key, V value) {
        return core.findBy(key, value);
    }

    @Override
    public <V> E getBy(Metamodel.Key<E, V> key, V value) {
        return core.getBy(key, value);
    }

    @Override
    public <V extends Data> Optional<E> findByRef(Metamodel.Key<E, V> key, Ref<V> value) {
        return core.findByRef(key, value);
    }

    @Override
    public <V extends Data> E getByRef(Metamodel.Key<E, V> key, Ref<V> value) {
        return core.getByRef(key, value);
    }

    // Page methods.

    @Override
    public Page<E> page(int pageNumber, int pageSize) {
        return core.page(pageNumber, pageSize);
    }

    @Override
    public Page<E> page(Pageable pageable) {
        return core.page(pageable);
    }

    @Override
    public Page<Ref<E>> pageRef(int pageNumber, int pageSize) {
        return core.pageRef(pageNumber, pageSize);
    }

    @Override
    public Page<Ref<E>> pageRef(Pageable pageable) {
        return core.pageRef(pageable);
    }

    @Override
    public List<E> findAll() {
        return core.findAll();
    }

    @Override
    public List<Ref<E>> findAllRef() {
        return selectRef().getResultList();
    }

    @Override
    public List<E> findAllById(Iterable<ID> ids) {
        return core.findAllById(ids);
    }

    @Override
    public List<E> findAllByRef(Iterable<Ref<E>> refs) {
        return core.findAllByRef(refs);
    }

    @Override
    public void insert(Iterable<E> entities) {
        core.insert(entities);
    }

    @Override
    public void insert(Iterable<E> entities, boolean ignoreAutoGenerate) {
        core.insert(entities, ignoreAutoGenerate);
    }

    @Override
    public List<ID> insertAndFetchIds(Iterable<E> entities) {
        return core.insertAndFetchIds(entities);
    }

    @Override
    public List<E> insertAndFetch(Iterable<E> entities) {
        return core.insertAndFetch(entities);
    }

    @Override
    public void update(Iterable<E> entities) {
        core.update(entities);
    }

    @Override
    public List<E> updateAndFetch(Iterable<E> entities) {
        return core.updateAndFetch(entities);
    }

    @Override
    public void upsert(Iterable<E> entities) {
        core.upsert(entities);
    }

    @Override
    public List<ID> upsertAndFetchIds(Iterable<E> entities) {
        return core.upsertAndFetchIds(entities);
    }

    @Override
    public List<E> upsertAndFetch(Iterable<E> entities) {
        return core.upsertAndFetch(entities);
    }

    @Override
    public void remove(Iterable<E> entities) {
        core.remove(entities);
    }

    @Override
    public void removeByRef(Iterable<Ref<E>> refs) {
        core.removeByRef(refs);
    }

    @Override
    public long countById(Stream<ID> ids) {
        return core.countById(ids);
    }

    @Override
    public long countById(Stream<ID> ids, int chunkSize) {
        return core.countById(ids, chunkSize);
    }

    @Override
    public long countByRef(Stream<Ref<E>> refs) {
        return core.countByRef(refs);
    }

    @Override
    public long countByRef(Stream<Ref<E>> refs, int chunkSize) {
        return core.countByRef(refs, chunkSize);
    }

    @Override
    public void insert(Stream<E> entities) {
        core.insert(entities);
    }

    @Override
    public void insert(Stream<E> entities, boolean ignoreAutoGenerate) {
        core.insert(entities, ignoreAutoGenerate);
    }

    @Override
    public void insert(Stream<E> entities, int batchSize) {
        core.insert(entities, batchSize);
    }

    @Override
    public void insert(Stream<E> entities, int batchSize, boolean ignoreAutoGenerate) {
        core.insert(entities, batchSize, ignoreAutoGenerate);
    }

    @Override
    public void update(Stream<E> entities) {
        core.update(entities);
    }

    @Override
    public void update(Stream<E> entities, int batchSize) {
        core.update(entities, batchSize);
    }

    @Override
    public void upsert(Stream<E> entities) {
        core.upsert(entities);
    }

    @Override
    public void upsert(Stream<E> entities, int batchSize) {
        core.upsert(entities, batchSize);
    }

    @Override
    public void remove(Stream<E> entities) {
        core.remove(entities);
    }

    @Override
    public void remove(Stream<E> entities, int batchSize) {
        core.remove(entities, batchSize);
    }

    @Override
    public void removeByRef(Stream<Ref<E>> refs) {
        core.removeByRef(refs);
    }

    @Override
    public void removeByRef(Stream<Ref<E>> refs, int batchSize) {
        core.removeByRef(refs, batchSize);
    }

    @Override
    public ORMTemplate orm() {
        return new ORMTemplateImpl(core.orm());
    }
}

File: storm-kotlin/src/main/kotlin/st/orm/repository/impl/EntityRepositoryImpl.kt
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.repository.impl

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.fold
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.stream.consumeAsFlow
import st.orm.Data
import st.orm.Entity
import st.orm.Ref
import st.orm.repository.EntityRepository
import st.orm.template.*
import st.orm.template.impl.ModelImpl
import st.orm.template.impl.ORMTemplateImpl
import st.orm.template.impl.QueryBuilderImpl
import kotlin.reflect.KClass

/**
 */
class EntityRepositoryImpl<E, ID : Any>(
    private val core: st.orm.core.repository.EntityRepository<E, ID>
) : EntityRepository<E, ID> where E : Data, E : Entity<ID> {

    override val model: Model<E, ID>
        get() = ModelImpl(core.model())

    override val orm: ORMTemplate
        get() = ORMTemplateImpl(core.orm())

    override fun ref(id: ID): Ref<E> =
        core.ref(id)

    override fun ref(entity: E): Ref<E> =
        core.ref(entity)

    override fun unload(entity: E): Ref<E> =
        core.unload(entity)

    override fun select(): QueryBuilder<E, E, ID> =
        QueryBuilderImpl(core.select())

    override fun selectCount(): QueryBuilder<E, Long, ID> =
        QueryBuilderImpl(core.selectCount())

    override fun <R : Any> select(selectType: KClass<R>): QueryBuilder<E, R, ID> =
        QueryBuilderImpl(core.select(selectType.java))

    override fun selectRef(): QueryBuilder<E, Ref<E>, ID> =
        QueryBuilderImpl(core.selectRef())

    override fun <R : Any> select(
        selectType: KClass<R>,
        template: TemplateString
    ): QueryBuilder<E, R, ID> =
        QueryBuilderImpl(core.select(selectType.java, template.unwrap))

    override fun <R : Data> selectRef(refType: KClass<R>): QueryBuilder<E, Ref<R>, ID> =
        QueryBuilderImpl<E, Ref<R>, ID>(core.selectRef(refType.java))

    override fun delete(): QueryBuilder<E, *, ID> =
        QueryBuilderImpl(core.delete())

    override fun count(): Long =
        core.count()

    override fun exists(): Boolean =
        core.exists()

    override fun existsById(id: ID): Boolean =
        core.existsById(id)

    override fun existsByRef(ref: Ref<E>): Boolean =
        core.existsByRef(ref)

    override fun insert(entity: E) =
        core.insert(entity)

    override fun insert(entity: E, ignoreAutoGenerate: Boolean) =
        core.insert(entity, ignoreAutoGenerate)

    override fun insertAndFetchId(entity: E): ID =
        core.insertAndFetchId(entity)

    override fun insertAndFetch(entity: E): E =
        core.insertAndFetch(entity)

    override fun update(entity: E) =
        core.update(entity)

    override fun updateAndFetch(entity: E): E =
        core.updateAndFetch(entity)

    override fun upsert(entity: E) =
        core.upsert(entity)

    override fun upsertAndFetchId(entity: E): ID =
        core.upsertAndFetchId(entity)

    override fun upsertAndFetch(entity: E): E =
        core.upsertAndFetch(entity)

    override fun delete(entity: E) =
        core.delete(entity)

    override fun deleteById(id: ID) =
        core.deleteById(id)

    override fun deleteByRef(ref: Ref<E>) =
        core.deleteByRef(ref)

    override fun deleteAll() =
        core.deleteAll()

    override fun findById(id: ID): E? =
        core.findById(id).orElse(null)

    override fun findByRef(ref: Ref<E>): E? =
        core.findByRef(ref).orElse(null)

    override fun getById(id: ID): E =
        core.getById(id)

    override fun getByRef(ref: Ref<E>): E =
        core.getByRef(ref)

    override fun findAll(): List<E> =
        core.findAll()

    override fun findAllById(ids: Iterable<ID>): List<E> =
        core.findAllById(ids)

    override fun findAllByRef(refs: Iterable<Ref<E>>): List<E> =
        core.findAllByRef(refs)

    override fun insert(entities: Iterable<E>) =
        core.insert(entities)

    override fun insert(entities: Iterable<E>, ignoreAutoGenerate: Boolean) =
        core.insert(entities, ignoreAutoGenerate)

    override fun insertAndFetchIds(entities: Iterable<E>): List<ID> =
        core.insertAndFetchIds(entities)

    override fun insertAndFetch(entities: Iterable<E>): List<E> =
        core.insertAndFetch(entities)

    override fun update(entities: Iterable<E>) =
        core.update(entities)

    override fun updateAndFetch(entities: Iterable<E>): List<E> =
        core.updateAndFetch(entities)

    override fun upsert(entities: Iterable<E>) =
        core.upsert(entities)

    override fun upsertAndFetchIds(entities: Iterable<E>): List<ID> =
        core.upsertAndFetchIds(entities)

    override fun upsertAndFetch(entities: Iterable<E>): List<E> =
        core.upsertAndFetch(entities)

    override fun delete(entities: Iterable<E>) =
        core.delete(entities)

    override fun deleteByRef(refs: Iterable<Ref<E>>) =
        core.deleteByRef(refs)

    override fun selectAll(): Flow<E> =
        core.selectAll().consumeAsFlow()

    override fun selectById(ids: Flow<ID>): Flow<E> =
        ids.chunked(core.defaultChunkSize)
            .flatMapConcat { core.findAllById(it).asFlow() }

    override fun selectByRef(refs: Flow<Ref<E>>): Flow<E> =
        refs.chunked(core.defaultChunkSize)
            .flatMapConcat { core.findAllByRef(it).asFlow() }

    override fun selectById(ids: Flow<ID>, chunkSize: Int): Flow<E> =
        ids.chunked(chunkSize)
            .flatMapConcat { core.findAllById(it).asFlow() }

    override fun selectByRef(refs: Flow<Ref<E>>, chunkSize: Int): Flow<E> =
        refs.chunked(chunkSize)
            .flatMapConcat { core.findAllByRef(it).asFlow() }

    override suspend fun countById(ids: Flow<ID>): Long =
        ids.chunked(core.defaultChunkSize)
            .map { chunk -> core.countById(chunk.stream()) }
            .fold(0L) { acc, v -> acc + v }

    override suspend fun countById(ids: Flow<ID>, chunkSize: Int): Long =
        ids.chunked(chunkSize)
            .map { chunk -> core.countById(chunk.stream()) }
            .fold(0L) { acc, v -> acc + v }

    override suspend fun countByRef(refs: Flow<Ref<E>>): Long =
        refs.chunked(core.defaultChunkSize)
            .map { chunk -> core.countByRef(chunk.stream()) }
            .fold(0L) { acc, v -> acc + v }

    override suspend fun countByRef(refs: Flow<Ref<E>>, chunkSize: Int): Long =
        refs.chunked(chunkSize)
            .map { chunk -> core.countByRef(chunk.stream()) }
            .fold(0L) { acc, v -> acc + v }

    override suspend fun insert(entities: Flow<E>) =
        entities.chunked(core.defaultBatchSize)
            .collect { core.insert(it) }

    override suspend fun insert(entities: Flow<E>, ignoreAutoGenerate: Boolean) =
        entities.chunked(core.defaultBatchSize)
            .collect { core.insert(it, ignoreAutoGenerate) }

    override suspend fun insert(entities: Flow<E>, batchSize: Int) =
        entities.chunked(batchSize)
            .collect { core.insert(it) }

    override suspend fun insert(entities: Flow<E>, batchSize: Int, ignoreAutoGenerate: Boolean) =
        entities.chunked(batchSize)
            .collect { core.insert(it, ignoreAutoGenerate) }

    override fun insertAndFetchIds(entities: Flow<E>): Flow<ID> =
        entities.chunked(core.defaultBatchSize)
            .flatMapConcat { core.insertAndFetchIds(it).asFlow() }

    override fun insertAndFetch(entities: Flow<E>): Flow<E> =
        entities.chunked(core.defaultBatchSize)
            .flatMapConcat { core.insertAndFetch(it).asFlow() }

    override fun insertAndFetchIds(
        entities: Flow<E>,
        batchSize: Int
    ): Flow<ID> =
        entities.chunked(batchSize)
            .flatMapConcat { core.insertAndFetchIds(it).asFlow() }

    override fun insertAndFetch(entities: Flow<E>, batchSize: Int): Flow<E> =
        entities.chunked(batchSize)
            .flatMapConcat { core.insertAndFetch(it).asFlow() }

    override suspend fun update(entities: Flow<E>) =
        entities.chunked(core.defaultBatchSize)
            .collect { core.update(it) }

    override suspend fun update(entities: Flow<E>, batchSize: Int) =
        entities.chunked(batchSize)
            .collect { core.update(it) }

    override fun updateAndFetch(entities: Flow<E>): Flow<E> =
        entities.chunked(core.defaultBatchSize)
            .flatMapConcat { core.updateAndFetch(it).asFlow() }

    override fun updateAndFetch(entities: Flow<E>, batchSize: Int): Flow<E> =
        entities.chunked(batchSize)
            .flatMapConcat { core.updateAndFetch(it).asFlow() }

    override suspend fun upsert(entities: Flow<E>) =
        entities.chunked(core.defaultBatchSize)
            .collect { core.upsert(it) }

    override suspend fun upsert(entities: Flow<E>, batchSize: Int) =
        entities.chunked(batchSize)
            .collect { core.upsert(it) }

    override fun upsertAndFetchIds(entities: Flow<E>): Flow<ID> =
        entities.chunked(core.defaultBatchSize)
            .flatMapConcat { core.upsertAndFetchIds(it).asFlow() }

    override fun upsertAndFetch(entities: Flow<E>): Flow<E> =
        entities.chunked(core.defaultBatchSize)
            .flatMapConcat { core.upsertAndFetch(it).asFlow() }

    override fun upsertAndFetchIds(
        entities: Flow<E>,
        batchSize: Int
    ): Flow<ID> =
        entities.chunked(batchSize)
            .flatMapConcat { core.upsertAndFetchIds(it).asFlow() }

    override fun upsertAndFetch(entities: Flow<E>, batchSize: Int): Flow<E> =
        entities.chunked(batchSize)
            .flatMapConcat { core.upsertAndFetch(it).asFlow() }

    override suspend fun delete(entities: Flow<E>) =
        entities.chunked(core.defaultBatchSize)
            .collect { core.delete(it) }

    override suspend fun delete(entities: Flow<E>, batchSize: Int) =
        entities.chunked(batchSize)
            .collect { core.delete(it) }

    override suspend fun deleteByRef(refs: Flow<Ref<E>>) =
        refs.chunked(core.defaultBatchSize)
            .collect { core.deleteByRef(it) }

    override suspend fun deleteByRef(refs: Flow<Ref<E>>, batchSize: Int) =
        refs.chunked(batchSize)
            .collect { core.deleteByRef(it) }
}
